packages/babel-helpers/src/helpers/applyDecs.js

```diff
@@ -0,0 +1,728 @@
+        return metadataForKey.constructor;
+      }
+    },
+    setMetadata(key, value) {
+      if (typeof key !== "symbol") {
+        throw new TypeError("Metadata keys must be symbols, received: " + key);
+      }
+
+      var metadataForKey = metadataMap[key];
+
+      if (metadataForKey === void 0) {
+        metadataForKey = metadataMap[key] = {};
+      }
+
+      if (kind === 1 /* PUBLIC */) {
+        var pub = metadataForKey.public;
+
+        if (pub === void 0) {
+          pub = metadataForKey.public = Object.create(null);
+        }
+
+        pub[property] = value;
+      } else if (kind === 2 /* PRIVATE */) {
+        var priv = metadataForKey.priv;
+
+        if (priv === void 0) {
+          priv = metadataForKey.private = new Map();
+        }
+
+        priv.set(property, value);
+      } else {
+        metadataForKey.constructor = value;
+      }
+    },
+  };
+}
+
+function convertMetadataMapToFinal(obj, metadataMap) {
+  var parentMetadataMap = obj[Symbol.metadata];
+  var metadataKeys = Object.getOwnPropertySymbols(metadataMap);
+
+  if (metadataKeys.length === 0) return;
+
+  for (var i = 0; i < metadataKeys.length; i++) {
+    var key = metadataKeys[i];
+    var metaForKey = metadataMap[key];
+    var parentMetaForKey = parentMetadataMap ? parentMetadataMap[key] : null;
+
+    var pub = metaForKey.public;
+    var parentPub = parentMetaForKey ? parentMetaForKey.public : null;
+
+    if (pub && parentPub) {
+      Object.setPrototypeOf(pub, parentPub);
+    }
+
+    var priv = metaForKey.private;
+
+    if (priv) {
+      var privArr = Array.from(priv.values());
+      var parentPriv = parentMetaForKey ? parentMetaForKey.private : null;
+
+      if (parentPriv) {
+        privArr = privArr.concat(parentPriv);
+      }
+
+      metaForKey.private = privArr;
+    }
+
+    if (parentMetaForKey) {
+      Object.setPrototypeOf(metaForKey, parentMetaForKey);
+    }
+  }
+
+  if (parentMetadataMap) {
+    Object.setPrototypeOf(metadataMap, parentMetadataMap);
+  }
+
+  obj[Symbol.metadata] = metadataMap;
+}
+
+function createAddInitializerMethod(initializers) {
+  return function addInitializer(initializer) {
+    assertValidInitializer(initializer);
+    initializers.push(initializer);
+  };
+}
+
+function memberDecCtx(
+  base,
+  name,
+  desc,
+  metadataMap,
+  initializers,
+  kind,
+  isStatic,
+  isPrivate
+) {
+  var kindStr;
+
+  switch (kind) {
+    case 1 /* ACCESSOR */:
+      kindStr = "accessor";
+      break;
+    case 2 /* METHOD */:
+      kindStr = "method";
+      break;
+    case 3 /* GETTER */:
+      kindStr = "getter";
+      break;
+    case 4 /* SETTER */:
+      kindStr = "setter";
+      break;
+    default:
+      kindStr = "field";
+  }
+
+  var ctx = {
+    kind: kindStr,
+    name: isPrivate ? "#" + name : name,
+    isStatic: isStatic,
+    isPrivate: isPrivate,
+  };
+
+  if (kind !== 0 /* FIELD */) {
+    ctx.addInitializer = createAddInitializerMethod(initializers);
```

Comment:
This might trigger a setter on `Object.prototype`, we can use `Object.defineProperty`.